File: backend/app/api/market_data.py
"""REST router — matches src/services/api/market-data.ts's expected surface."""
from datetime import date, datetime, timedelta, timezone
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import EtfFlow as EtfFlowModel, LargeOrderLevel as LargeOrderLevelModel
from app.db.session import get_db
from app.historical import service as historical_service
from app.schemas import CandlesRequest, CandlesResponse, EtfFlow, HealthResponse, LargeOrderLevel, Symbol, Tick
from app.store import store
from app.symbols import get_symbol, list_symbols

router = APIRouter(prefix="/api/market-data")


@router.get("/symbols", response_model=list[Symbol])
async def get_symbols() -> list[Symbol]:
    return [info.to_schema() for info in list_symbols()]


@router.post("/candles/{symbol}", response_model=CandlesResponse)
async def get_candles(symbol: str, body: CandlesRequest | None = None) -> CandlesResponse:
    symbol_info = get_symbol(symbol)
    if symbol_info is None:
        raise HTTPException(status_code=404, detail=f"unknown symbol: {symbol}")

    body = body or CandlesRequest()
    candles, metadata = await historical_service.get_candles(
        symbol_info,
        timeframe=body.timeframe,
        limit=body.limit,
        from_ms=body.from_,
        to_ms=body.to,
    )
    return CandlesResponse(candles=candles, metadata=metadata)


@router.get("/ticks/{symbol}", response_model=Tick)
async def get_tick(symbol: str) -> Tick:
    if get_symbol(symbol) is None:
        raise HTTPException(status_code=404, detail=f"unknown symbol: {symbol}")
    tick = store.get_tick(symbol)
    if tick is None:
        raise HTTPException(status_code=404, detail=f"no tick yet for symbol: {symbol}")
    return tick


@router.get("/ticks", response_model=dict[str, Tick])
async def get_ticks() -> dict[str, Tick]:
    return store.get_all_ticks()


@router.get("/health", response_model=HealthResponse)
async def get_health() -> HealthResponse:
    return HealthResponse(binance=store.get_health("binance"), okx=store.get_health("okx"))


@router.get("/etf-flows", response_model=list[EtfFlow])
async def get_etf_flows(
    from_: date | None = Query(default=None, alias="from"),
    to: date | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
) -> list[EtfFlow]:
    stmt = select(EtfFlowModel).order_by(EtfFlowModel.flow_date)
    if from_ is not None:
        stmt = stmt.where(EtfFlowModel.flow_date >= from_)
    if to is not None:
        stmt = stmt.where(EtfFlowModel.flow_date <= to)
    rows = (await db.execute(stmt)).scalars().all()
    return [
        EtfFlow(
            flowDate=row.flow_date,
            totalNetFlow=float(row.total_net_flow),
            observedAt=row.observed_at,
            updatedAt=row.updated_at,
        )
        for row in rows
    ]


@router.get("/large-order-book/history", response_model=list[LargeOrderLevel])
async def get_large_order_book_history(
    base: Literal["BTC", "ETH"],
    source: list[Literal["binance", "okx"]] = Query(default=["binance", "okx"]),
    threshold: int = Query(default=1_000_000, ge=0, le=10_000_000),
    from_: datetime | None = Query(default=None, alias="from"),
    to: datetime | None = Query(default=None),
    limit: int = Query(default=1000, ge=1, le=1000),
    db: AsyncSession = Depends(get_db),
) -> list[LargeOrderLevel]:
    now = datetime.now(timezone.utc)
    start = max(from_ or now - timedelta(days=30), now - timedelta(days=30))
    end = to or now
    symbols = [f"{exchange.upper()}:{base}USD" for exchange in source]
    stmt = (
        select(LargeOrderLevelModel)
        .where(
            LargeOrderLevelModel.symbol.in_(symbols),
            LargeOrderLevelModel.source.in_(source),
            LargeOrderLevelModel.peak_notional >= threshold,
            LargeOrderLevelModel.first_seen <= end,
            (LargeOrderLevelModel.ended_at.is_(None) | (LargeOrderLevelModel.ended_at >= start)),
        )
    )
    stmt = stmt.order_by(LargeOrderLevelModel.first_seen.desc()).limit(limit)
    try:
        rows = (await db.execute(stmt)).scalars().all()
    except Exception as exc:
        raise HTTPException(status_code=503, detail="large-order history unavailable") from exc
    return [
        LargeOrderLevel(
            id=str(row.id),
            source=row.source,
            symbol=row.symbol,
            side=row.side,
            price=float(row.price),
            quantity=float(row.quantity),
            currentNotional=float(row.current_notional),
            peakNotional=float(row.peak_notional),
            firstSeen=row.first_seen,
            lastSeen=row.last_seen,
            endedAt=row.ended_at,
        )
        for row in rows
    ]

File: src/services/api/market-data.ts
import { request } from "./request";

export interface MarketDataSymbol {
  id: string;
  name: string;
  displayName: string | null;
  category: string;
  contractSize: number;
  tickSize: number;
  tickValue: number;
  marginPercent: number;
  maxLeverage: number;
  commission: number;
  swapLong: number;
  swapShort: number;
  tradingHoursStart: string | null;
  tradingHoursEnd: string | null;
  isActive: boolean;
  // Optional/extra fields backend may include
  symbol?: string;
  digits?: number;
  pipDigits?: number;
  [key: string]: unknown;
}

// Mirrors backend/app/schemas.py's InstrumentSpec — real per-instrument
// specs from OKX's public instruments endpoint, used by Position Builder.
export interface MarketDataInstrumentSpec {
  instId: string;
  instType: string;
  ctVal: number;
  ctValCcy: string;
  lotSz: number;
  minSz: number;
  tickSz: number;
  settleCcy: string;
  quoteCcy: string;
  baseCcy: string;
  maxLever: number;
}

export interface MarketDataCandle {
  time: number;
  timestamp: number | string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface MarketDataCandleMetadata {
  historicalCoverageStart: number | null;
  isPartial: boolean;
  backfillQueued: boolean;
}

// Mirrors backend/app/schemas.py ExchangeHealth / HealthResponse — one entry
// per exchange, keyed by exchange name.
export interface MarketDataExchangeHealth {
  connected: boolean;
  /** Any feed traffic — ticker, candles or book. Connection liveness. */
  lastEventAt: number | null; // unix ms
  /**
   * Ticker events only: the last time a *price* actually arrived. Judge
   * staleness on this, never on lastEventAt — a dead ticker channel leaves
   * prices frozen while candle and book traffic keep lastEventAt current.
   * Null until the first tick of a feed's lifetime.
   */
  lastTickAt: number | null; // unix ms
}

export type MarketDataHealth = Record<string, MarketDataExchangeHealth>;

export interface MarketDataCandlesPayload {
  candles: MarketDataCandle[];
  metadata: MarketDataCandleMetadata;
}

export interface MarketDataTick {
  symbol: string;
  bid: number;
  ask: number;
  timestamp?: number | string;
}

export interface EtfFlow {
  flowDate: string;
  totalNetFlow: number;
  observedAt: string | null;
  updatedAt: string;
}

export type LargeOrderSource = "binance" | "okx";
export type LargeOrderThreshold = 0 | 500_000 | 1_000_000 | 3_000_000 | 5_000_000 | 10_000_000;

export interface LargeOrderLevel {
  id: string;
  source: LargeOrderSource;
  symbol: string;
  side: "bid" | "ask";
  price: number;
  quantity: number;
  currentNotional: number;
  peakNotional: number;
  firstSeen: string;
  lastSeen: string;
  endedAt: string | null;
}

export interface EconomicEvent {
  id: string;
  time: string;
  currency: string;
  event: string;
  impact: "low" | "medium" | "high";
  forecast?: string;
  previous?: string;
  actual?: string;
  country: string;
  [key: string]: unknown;
}

export const marketdataApi = {
  // ── Market Data ──
  getSymbols: () => request<MarketDataSymbol[]>("/market-data/symbols"),

  getInstrument: (symbol: string) =>
    request<MarketDataInstrumentSpec>(`/market-data/instrument/${encodeURIComponent(symbol)}`),

  getTick: (symbol: string) => request<MarketDataTick>(`/market-data/ticks/${symbol}`),

  getTicks: () => request<Record<string, MarketDataTick>>("/market-data/ticks"),

  getCandles: (
    symbol: string,
    timeframe: string,
    limit?: number,
    range?: { fromMs: number; toMs: number },
  ) => {
    const payload: Record<string, number | string> = { timeframe };
    if (limit != null) payload.limit = limit;
    if (range) {
      payload.from = range.fromMs;
      payload.to = range.toMs;
    }
    return request<MarketDataCandle[] | MarketDataCandlesPayload>(
      `/market-data/candles/${symbol}`,
      {
        method: "POST",
        body: JSON.stringify(payload),
      },
    ).then((res) => (Array.isArray(res) ? res : res.candles));
  },

  getCandlesWithMeta: (
    symbol: string,
    timeframe: string,
    limit?: number,
    range?: { fromMs: number; toMs: number },
  ): Promise<MarketDataCandlesPayload> => {
    const payload: Record<string, number | string> = { timeframe };
    if (limit != null) payload.limit = limit;
    if (range) {
      payload.from = range.fromMs;
      payload.to = range.toMs;
    }
    return request<MarketDataCandle[] | MarketDataCandlesPayload>(
      `/market-data/candles/${symbol}`,
      {
        method: "POST",
        body: JSON.stringify(payload),
      },
    ).then((res) =>
      Array.isArray(res)
        ? {
            candles: res,
            metadata: { isPartial: false, backfillQueued: false, historicalCoverageStart: null },
          }
        : res,
    );
  },

  getMarketDataHealth: () => request<MarketDataHealth>("/market-data/health"),

  getEconomicCalendar: (currencies?: string[], from?: string, to?: string) => {
    const params = new URLSearchParams();
    if (currencies?.length) params.set("currencies", currencies.join(","));
    if (from) params.set("from", from);
    if (to) params.set("to", to);
    return request<EconomicEvent[]>(`/market-data/economic-calendar?${params.toString()}`);
  },

  getMarketDataStaleness: () => request<Record<string, unknown>>("/market-data/staleness"),

  getEtfFlows: (from?: string, to?: string) => {
    const params = new URLSearchParams();
    if (from) params.set("from", from);
    if (to) params.set("to", to);
    const qs = params.toString();
    return request<EtfFlow[]>(`/market-data/etf-flows${qs ? `?${qs}` : ""}`);
  },

  getLargeOrderBookHistory: (
    base: "BTC" | "ETH",
    sources: LargeOrderSource[],
    threshold: LargeOrderThreshold,
    from?: string,
    to?: string,
    limit?: number,
  ) => {
    const params = new URLSearchParams({ base, threshold: String(threshold) });
    sources.forEach((source) => params.append("source", source));
    if (from) params.set("from", from);
    if (to) params.set("to", to);
    if (limit) params.set("limit", String(limit));
    return request<LargeOrderLevel[]>(`/market-data/large-order-book/history?${params}`);
  },
};
